### PR TITLE
copr: add sub_datetime_and_string

### PR DESCRIPTION
**Pull request by Jibbow**
Hi there :wave: 
This is my first PR to the TiKV in an attempt to apply for the LFX Mentorship Program.
I am happy about all kind of feedback. Thank you! :blush: 

### What problem does this PR solve?
Implements `sub_datetime_and_string()` for coprocessor.

### Check List #5751, #9016 

Tests 

- Unit test

### Release note 

- coprocessor supports `sub_datetime_and_string()`

### CODE REVIEW

**Review by Xuanwo — APPROVED** [review]

**Comment by Xuanwo** [comment]
cc @andylokandy

File: components/tidb_query_expr/src/impl_time.rs [hunk 1]
```diff
@@ -382,6 +382,33 @@ pub fn sub_datetime_and_duration(
...
+        Err(_) => return Ok(None),
+    };
+
+    let res = match datetime.checked_sub(ctx, duration_str) {
+        Some(res) => res,
+        None => {
+            return ctx
+                .handle_invalid_time_error(Error::overflow(
+                    "DATETIME",
+                    format!("({} + {})", datetime, duration_str),
```
Comment (andylokandy) [inline components/tidb_query_expr/src/impl_time.rs:404]:
```suggestion
                    format!("({} - {})", datetime, duration_str),
```

**Review by andylokandy — COMMENTED** [review]
rest LGTM!

**Review by andylokandy — COMMENTED** [review]
rest LGTM!

File: components/tidb_query_expr/src/impl_time.rs [hunk 2]
```diff
@@ -382,6 +382,33 @@ pub fn sub_datetime_and_duration(
...
 
+#[rpn_fn(capture = [ctx])]
+#[inline]
+pub fn sub_datetime_and_string(
+    ctx: &mut EvalContext,
+    datetime: &DateTime,
+    duration_str: BytesRef,
+) -> Result<Option<DateTime>> {
+    let duration_str = std::str::from_utf8(&duration_str).map_err(Error::Encoding)?;
+    let duration_str = match Duration::parse(ctx, duration_str, MAX_FSP) {
```
Comment (andylokandy) [inline components/tidb_query_expr/src/impl_time.rs:393]:
```suggestion
    let duration = match Duration::parse(ctx, duration_str, MAX_FSP) {
```

File: components/tidb_query_expr/src/impl_time.rs [hunk 3]
```diff
@@ -382,6 +382,33 @@ pub fn sub_datetime_and_duration(
...
+        Ok(arg) => arg,
```
Comment (andylokandy) [inline components/tidb_query_expr/src/impl_time.rs:394]:
```suggestion
        Ok(duration) => duration,
```

**Review by andylokandy — APPROVED** [review]

File: components/tidb_query_expr/src/impl_time.rs [hunk 4]
```diff
@@ -382,6 +382,33 @@ pub fn sub_datetime_and_duration(
...
+    datetime: &DateTime,
+    duration_str: BytesRef,
+) -> Result<Option<DateTime>> {
+    let duration_str = std::str::from_utf8(&duration_str).map_err(Error::Encoding)?;
+    let duration = match Duration::parse(ctx, duration_str, MAX_FSP) {
+        Ok(duration) => duration,
+        Err(_) => return Ok(None),
+    };
+
+    let res = match datetime.checked_sub(ctx, duration_str) {
```
Comment (andylokandy) [inline components/tidb_query_expr/src/impl_time.rs:398]: this needs to rename

File: components/tidb_query_expr/src/impl_time.rs [hunk 5]
```diff
@@ -382,6 +382,33 @@ pub fn sub_datetime_and_duration(
...
+        Some(res) => res,
+        None => {
+            return ctx
+                .handle_invalid_time_error(Error::overflow(
+                    "DATETIME",
+                    format!("({} - {})", datetime, duration_str),
```
Comment (andylokandy) [inline components/tidb_query_expr/src/impl_time.rs:404]: ditto

**Pull request merged**